src/main/java/net/logstash/logback/util/ReusableJsonFormatterPool.java

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright 2013-2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package net.logstash.logback.util;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.lang.ref.SoftReference;
+import java.util.Deque;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentLinkedDeque;
+
+import net.logstash.logback.composite.CompositeJsonFormatter;
+
+import ch.qos.logback.core.spi.DeferredProcessingAware;
+
+/**
+ * @author brenuart
+ *
+ */
+public class ReusableJsonFormatterPool<Event extends DeferredProcessingAware> {
+
+    private final int minBufferSize;
+    
+    private final CompositeJsonFormatter<Event> formatterFactory;
+    
+    private SoftReference<Deque<ReusableJsonFormatter>> formatters = new SoftReference<>(null);
+    
+
+    public ReusableJsonFormatterPool(CompositeJsonFormatter<Event> formatterFactory, int minBufferSize) {
+        this.formatterFactory = Objects.requireNonNull(formatterFactory);
+        this.minBufferSize = minBufferSize;
+    }
+
+    public class ReusableJsonFormatter implements Closeable {
+        private final ReusableByteBuffer buffer;
+        private final CompositeJsonFormatter<Event>.JsonFormatter formatter;
+        
+        ReusableJsonFormatter(ReusableByteBuffer buffer, CompositeJsonFormatter<Event>.JsonFormatter formatter) {
+            this.buffer = buffer;
+            this.formatter = formatter;
+        }
+        
+        public ReusableByteBuffer getBuffer() {
+            return buffer;
+        }
+        
+        public void write(Event event) throws IOException {
+            this.formatter.writeEvent(event);
+        }
+        
+        @Override
+        protected void finalize() throws Throwable {
+            dispose();
+        }
+        
+        public void dispose() throws IOException {
+            this.formatter.dispose();
+            this.buffer.reset();
+        }
+        
+        public void reset() {
+            this.buffer.reset();
+        }
+        
+        @Override
+        public void close() throws IOException {
+            release(this);
+        }
```

Comment:
It is unclear when users of this class should call dispose()/close()/reset().  Please document them to make it clear.

Also, can the visibility of dispose() and reset() be reduced?  I didn't see where they are used outside of the pool.

Reply:
Indeed... I'll clean this up - was still a _wip_ to get your feedback ;-)